### ISSUE TITLE
Copy Constructor for Search Listeners

### ISSUE THREAD

**Issue by VoglSebastian** [VoglSebastian/evosuite#15]
### Overview

Search Listener has a genetic Type parameter, that defines the Type of the observed Genetic algorithm.

A TestSuiteAdapter needs to map those Stopping conditions from SearchListener<TestSuiteChromosome> to SearchListener<TestChromosome> and preserve the current state.

A possibility to add such a mapping would be to add a copy Constructor (that is independent of the generic type) to all Search Listeners.

e.g:
```
public MaxGenerationStoppingCondition(MaxGenerationStoppingCondition<?> other) {
	this.currentIteration = other.currentIteration;
	this.maxIterations = other.maxIterations;
}
```

The advantage of this solution would be, that the logic of copying the state of the SearchListener, would be inside the class and the TestSuiteAdapter only has to call the copy constructor (in contrast to create a new Listener and copy the state with Getter and Setter Methods).

The main drawback is, that the changes have to be made in a lot of classes:

- [x] BranchCoverageMap *(singleton)*
- [x] DiversityObserver
- [x] FitnessLogger
- [x] FunctionalSearchListener
- [x] GlobalTimeStoppingCondition
- [x] LocalSearchBudget *(singleton)*
- [x] MaxFitnessEvaluationsStoppingCondition
- [x] MaxGenerationStoppingCondition
- [x] MaxLengthStoppingCondition
- [x] MaxStatementsStoppingCondition
- [x] MaxTestsStoppingCondition
- [x] MaxTimeStoppingCondition
- [x] MutationTestPool *(only contains static fields)*
- [x] MutationTimeoutStoppingCondition
- [x] ProgressMonitor
- [x] RMIStoppingCondition *(singleton)*
- [x] RelativeSuiteLengthBloatControl
- [x] RelativeTestLengthBloatControl
- [x] ResourceController
- [x] ShutdownTestWriter
- [x] SocketStoppingCondition
- [x] StatisticsListener
- [x] StoppingCondition *(interface)*
- [x] StoppingConditionImpl *(abstract class)*
- [x] SuiteFitnessEvaluationListener
- [x] TestCaseRecycler *(singleton)*
- [x] TimeDeltaStoppingCondition
- [x] ZeroFitnessStoppingCondition

### Contributing
Please push your changes to [`15-copy-constructor-for-search-listeners`](https://github.com/VoglSebastian/evosuite/tree/15-copy-constructor-for-search-listeners)

**Comment by schweikl** [comment]
Sounds reasonable, I'm in favor of this. This would simplify the implementation of #5

**Comment by VoglSebastian** [comment]
I just have seen, that the same issue is present for PopulationLimit and have already fixed it.

(Commit 96f445b)

**Comment by schweikl** [comment]
That's great :+1: Does that relate to #5? If so, could you mark the progress by checking the appropriate boxes?

**Comment by VoglSebastian** [comment]
It does relate to #5, but the method was already implemented, i just did a small refactoring, to have similar approaches to the search listener.

So population limits are now copied with the copy constructor instead of creating a new instance.

**Comment by schweikl** [comment]
`ProgressMonitor` contains a `StoppingCondition` field. Every `StoppingCondition` is also a `SearchListener`, so it will have a copy constructor added in the course of fixing this issue. But invoking the copy constructor is only possible for concrete types and not interface types... Hence, shall we add a method `StoppingCondition<T> clone()` to the `StoppingCondition` interface, and every subclass implements it by calling its copy constructor with `this`? This way, the copy constructor of `ProgressMonitor`  could invoke the copy constructor of a `StoppingCondition` indirectly via its `clone()` method.

**Comment by schweikl** [comment]
Update: I have just tried my own suggestion and it seems to work fine. I have already pushed these changes to `15-copy-constructor-for-search-listeners`.

**Comment by schweikl** [comment]
Fixed by #22
